src/drivers/spi/spi.c

```diff
@@ -0,0 +1,81 @@
+/**
+ * @file spi.c
+ * @brief NOTE: When you use spi.core without spi.devfs
+ * you should make sure device-specific stuff is initialized
+ * somehow else
+ *
+ * @author Denis Deryugin <deryugin.denis@gmail.com>
+ * @version
+ * @date 03.12.2018
+ */
+
+#include <drivers/spi.h>
+#include <util/log.h>
+
+#include <assert.h>
+#include <errno.h>
+
+#include <embox/unit.h>
+#include <framework/mod/options.h>
+
+/* Note: it's array of pointers, not structures as usual */
+ARRAY_SPREAD_DEF(struct spi_device *, spi_device_registry);
+
+static int spi_init(void) {
+	struct spi_device **dev;
+
+	array_spread_foreach_ptr(dev, spi_device_registry) {
+		if ((*dev)->spi_ops->init) {
+			(*dev)->spi_ops->init(*dev);
+		} else {
+			log_warning("SPI%d has no init funcion", (*dev)->id);
+		}
+	}
+
+	return 0;
+}
+
+EMBOX_UNIT_INIT(spi_init);
+
+struct spi_device *spi_dev_by_id(int id) {
```

Comment:
`spi_dev_by_id` searches for device without comparing with its spi_device->id, as I can see. But what if I have two spi devices with id's 0 and 1, but they were added to spi_device_registry in the reverse order somehow (SPI_DEF(spi1), then SPI_DEF(spi0))? Then when I write spi_dev_by_id(0), it's possible situation when I can get second device, not the first one?

Reply:
But how I can make sure `spi_dev_by_id` returned _right_ spi controller, since the enumeration within ARRAY_SPREAD depends on linker?

Reply:
@alexkalmuk  It's not so significant now
@Deryugin In i2c I set number of i2c controller (bus) statically, and during i2c module init check are the numbers not cross between each other. It's rather useful to understand which real controller used

Reply:
Check out last commits pls